### PR TITLE
Simplify interface and add support for ContT

### PR DESCRIPTION
**Pull request by EduardSergeev**
* Add `runFSMCont` compatible with `ContT` which does not implement `MonadCatch`
* `transition` is now `Action` which is `MonadReader` and `MonadWriter`
* Incapsulate `PropertyM` in `Action` to simplify user's test code (safe on `lift`s)
* Add various combinators for defining transition and asserts
* `runFSM` now accepts `Gen (Transition e s m)` instead of `Gen [Transition e s m]` - the list of transition is created internally, it is "sized" and consists of only "valid" transitions for the current state, so no transition are discarded by QC during the test run - this way there is less chances that test fails due QC' "give up" (not enough valid test cases) which is used to be a problem before.

This interface should be easier to use than https://github.com/ambiata/disorder.hs/pull/40 (they are similar) so I am closing the previous PR.

### CODE REVIEW

**Comment by EduardSergeev** [comment]
Example of FSM testing for `sanction`: https://github.com/ambiata/sanction/pull/119

File: disorder-fsm/src/Disorder/FSM.hs [hunk 1]
```diff
@@ -1,50 +1,87 @@
...
 
 import           Test.QuickCheck.Gen
-import           Test.QuickCheck.Monadic
-import           Test.QuickCheck.Property
 
 
 -- | Defines a transition from state to state
 data Transition e s m = MkTransition {
     -- | Display name (used in case of failure)
     name :: String
```
Comment (etorreborre) [inline disorder-fsm/src/Disorder/FSM.hs]: Why `String` and not `Text`?

Comment (EduardSergeev) [inline disorder-fsm/src/Disorder/FSM.hs]: @etorreborre `name` is only used here in `Show` and `fail` both of which require `String`. Our `Text` would always be converted to `String` anyway.

Comment (etorreborre) [inline disorder-fsm/src/Disorder/FSM.hs]: Ok, and I suppose we will use literal strings 90% of the time anyway.

**Comment by nhibberd** [comment]
@EduardSergeev Is this ready to be merged?

**Comment by EduardSergeev** [comment]
@nhibberd Yes, once rebase it could be merged. I haven't done it yet because it still have this problem: https://github.com/ambiata/disorder.hs/issues/48 (still unresolved)

**Comment by markhibberd** [comment]
@EduardSergeev Is this worth rebase+merging in current state, and doing #48 in separate PR?

**Comment by EduardSergeev** [comment]
The [solution](https://github.com/ambiata/disorder.hs/pull/55) for #48 changes the interface defined here quite dramatically.
Since https://github.com/ambiata/disorder.hs/pull/55 is pretty much done (just some polishing to be done, as always, but I am quite happy with the current interface/implementation) so I think I can merge just https://github.com/ambiata/disorder.hs/pull/55 and close this one.

**Pull request closed without merging**